Avoid py token usage in the common path for circuit_to_dag

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adjusts the Python usage in the DAGCircuit::from_circuit_data code paths. Most of the data model is in Rust at this point with the exception of things explicitly defined in Python (ie a custom gate in Python), ParameterExpression, and control flow. While the latter two are in progress (see #13267 and #13271) we don't need to block the development of C API features that depend on circuit->dag conversion. The Python usage in this code path is around parameterized global phase and deepcopying custom python gates. Neither of these are possible from C because they're in Python and not exposed to C. This commit removes the py token from the function signatures and actively acquires the gil inline if we're using Python parts of the data model when they're needed.

### Details and comments

mtreinish committed:
Avoid py token usage in the common path for circuit_to_data

This commit adjusts the Python usage in the
DAGCircuit::from_circuit_data code paths. Most of the data model is in
Rust at this point with the exception of things explicitly defined in
Python (ie a custom gate in Python), ParameterExpression, and control
flow. While the latter two are in progress (see #13267 and #13271) we
don't need to block the development of C API features that depend on
circuit->dag conversion. The Python usage in this code path is around
parameterized global phase and deepcopying custom python gates.
Neither of these are possible from C because they're in Python and not
exposed to C. This commit removes the py token from the function
signatures and actively acquires the gil inline if we're using Python
parts of the data model when they're needed.

kevinhartman reviewed (dismissed):
This looks good to me (... and, it's a lot more minimal than I was expecting given your description heh 🙂).

jakelishman commented on crates/circuit/src/packed_instruction.rs:
```diff
@@ -452,42 +452,46 @@ impl PackedOperation {
-        memo: Option<&Bound<'py, PyDict>>,
-    ) -> PyResult<Self> {
-        let deepcopy = DEEPCOPY.get_bound(py);
+    pub fn py_deepcopy(&self, memo: Option<&Bound<'_, PyDict>>) -> PyResult<Self> {
```
Imo this change is incorrect.  All callers of `py_deepcopy` should manage retrieving the GIL themselves - if you're asking for this method, you _must_ be GIL-aware yourself; there's no need to call it if you're not because you can just use `clone`.

Maybe more have uses have been added, but my recollection is that all uses of this are conditional on a `copy_operations: bool` flag, which can be used to conditionally gate whether the GIL is acquired before this function is called.

mtreinish replied:
All the usage is conditional on a boolean for either "deepcopy" or copy operations. The thing which is problematic is that this method is not limited to python space objects. It's defined on `PackedInstruction` which almost certainly does not have to be a Python object under the covers.

I'm fine in principle unrolling this a layer and just never setting the appropriate boolean in a C context. But I think this case we should not have a deepcopy method on `PackedInstruction` as it's leaking python into a core object where it doesn't need to be.

mtreinish replied:
I took care of this in: https://github.com/Qiskit/qiskit/pull/14589/commits/2e496e60e834b4a54af06760b080758754f9db1d I removed the method on packed instruction and made the py copy methods part of a trait shared by the python defined operation structs. This makes the deepcopy explicitly only something that is a concern when working with python objects and it meets both our requirements here.

jakelishman commented:
I removed this from the queue on the basis of the previous comment.  I don't mind being overruled and it re-added to the queue, but I think it should be considered first.

I think we should be being a lot more careful about "hide the py".  A method which is quite literally asking for Python-space deepcopies in its very name really feels like it _should_ require the GIL to call.

jakelishman commented:
Also, is the milestone set right?  I'd have assumed this was for 2.2.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into hide-the-py-circ-to-dag

mtreinish committed:
Move deepcopy Python usage up to Python defined types

In the previous commit the py token acquisition was moved down into the
PackedInstruction::py_deepcopy() method instead of being the
responsibility of the caller. This decision was made because we only
need the py token in a context where objects defined in Python are
encountered. The goal of hiding the py token is that in a context
without python at all (i.e. the standalone C API) nothing callable from
a pure rust context requires the py token. The token was moved down so
that we only ever try to acquire the GIL in contexts where we have a
python defined object. The method only ever uses it to deep copy in the
case of PackedInstruction containing a Python defined object.

In code review the concern was raised that deepcopy is a purely python
construct and the `py_deepcopy` should very explicitly be the caller's
responsibility to assert they have the GIL by passing the py token. To
address this concern but also enable a Python free path through circuit
to dag conversion this commit moves the `deepcopy()` (and `py_copy()`)
methods up a layer so that they're defined on the Python owned
operations. A new trait `PythonOperation` is added that contains these
methods and the trait is implemented for the Python defined operation
types: `PyGate`, `PyInstruction`, and `PyOperation`. This makes it more
explicit where these methods are callable and makes the python usage
explicit from the caller.

raynelfss reviewed (approved):
This looks good to me!

I thought about suggesting all of the matching to happen within a method or something similar, but with how it looks right now, I don't think there's a way around it considering that we will only call this for certain types of operations. We can always follow it up and think of a better-looking way of doing it.

That said, make sure you update the PRs description to include the `PythonOperation` trait's description and why it was added :)

mtreinish commented:
> That said, make sure you update the PRs description to include the `PythonOperation` trait's description and why it was added :)

It's in the commit message for the change: https://github.com/Qiskit/qiskit/pull/14589/commits/2e496e60e834b4a54af06760b080758754f9db1d
